### PR TITLE
work on merging the cmake changes

### PR DESCRIPTION
**Pull request by schweitzpgi**
The check-flang target doesn't work any longer, so we'll need to fix that before merging this PR.

This is a merge of Pat's PR on top of mono.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,258 +1,123 @@
...
-    # DLL platform -- put DLLs into bin.
-    set(LLVM_SHLIB_OUTPUT_INTDIR ${LLVM_RUNTIME_OUTPUT_INTDIR})
-  else()
-    set(LLVM_SHLIB_OUTPUT_INTDIR ${LLVM_LIBRARY_OUTPUT_INTDIR})
-  endif()
+  include(CMakeParseArguments)
+  include(AddLLVM)
+  include(HandleLLVMOptions)  
+  include(VersionFromVCS)
+  include(AddMLIR)
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply CMakeLists.txt]: OK. I'll make that change.

**Comment by schweitzpgi** [comment]
> Apart from my comment, my gcc out-of-tree build went well, the clang one is failing with a crazy issue:
> 
> ```
> [ 67%] Linking CXX executable ../../bin/tco
> /lib/../lib64/crt1.o: In function `_start':
> (.text+0x20): undefined reference to `main'
> clang-8: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [bin/tco] Error 1
> make[1]: *** [tools/tco/CMakeFiles/tco.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```
> 
> That must be something broken on my side. Investigating.

When I tried out-of-tree, I did rebuild my LLVM+MLIR binaries from the mono branch beforehand. That shouldn't make a difference, but ...

File: test/Runtime/CMakeLists.txt [hunk 2]
```diff
@@ -1,39 +1,15 @@
-#===-- test/Runtime/CMakeLists.txt -----------------------------------------===#
```
Comment (sscalpone) [inline test/Runtime/CMakeLists.txt]: Why remove?  Other projects include the LLVM license header in CMakeList.txt files?

Comment (schweitzpgi) [reply test/Runtime/CMakeLists.txt]: For what it is worth, the bulk of these changes come from Pat's change.

The assertion that LLVM CMakeLists.txt files include the license banner is false. So I assume that is why Pat has removed them from most of the files he touched in his PR.

Comment (schweitzpgi) [reply test/Runtime/CMakeLists.txt]: I basically just merged all of his changes and then went through and resolved comments and fixed major problems to this point. I still did not fix the lit test problem. But I need to do that ASAP.

**Pull request merged**